### PR TITLE
Support Nested Fragments

### LINKED ISSUE #170: Nested Fragments don't render
This doesn't render anything:
```jsx
<>
  <>
    text
  </>
<>
```

This works as expected
```jsx
<>
  <span>
    text
  </span>
<>
```

### PR DESCRIPTION
**Pull request by ricky-sb**
### Overview
Nested fragments with invalid markup aren't supported. So, for example, this won't work:
```jsx
<>
  <>
    text
  </>
<>
```

But this *does* work in React. Part of the issue is that when fragments don't contain valid markup, React will just happily insert it into the DOM anyway. So this is "valid" React code 😬. 

### Technical Dive

Basically we were [crashing here](https://github.com/near/bos-web-engine/blob/835bd57d9fd4a9f88b63a4183f95d5c019555295/packages/application/src/react.ts#L80-L89) because `children` did not contain any elements. 

### Example:
Valid `children`:

image

Invalid `children`:

image  

Notice the null type here.

## What's New
**For Valid Children**: If a fragment's children are all valid nodes (meaning, they have a type), they're returned directly. This keeps our DOM clean and avoids adding unnecessary wrappers.

**For Invalid Children**: Sometimes, you get fragments with text nodes or similar, which aren't valid on their own. Here, I introduced a small compromise: wrapping these in a `div`. It’s a bit of a sidestep from the ideal React purity, but it gets the job done without complicating the rendering logic too much.  

## Trade-offs
I pondered a bit on this. Ideally, we'd keep everything as pure to React's principles as possible, avoiding extra DOM elements. But, the reality of supporting a wider range of use-cases with minimal complexity led me down this path. It’s a bit of a trade-off:

- **Pros**: We're now more inclusive of different fragment structures, making this a bit more forgiving and flexible.
- **Cons**: The introduction of an extra `div` for handling invalid children in fragments isn't perfectly in line with React’s minimalist DOM philosophy, although we're already wrapping the root fragment anyway 🤷‍♂️ 

## Alternatives
I’m really keen to hear your thoughts on this. I think we could probably do something with the `parser.ts` and `react.ts` to support creating elements inside a fragment without wrapping it first, although I figured this was the simplest solution. 

Another option might be to say "we don't support invalid markup". Maybe we *shouldn't* support this in the first place!

Closes #170

### CODE REVIEW

**Comment by andy-haynes** [comment]
@ricky-sb thank you for your contribution and thoughtful writeup!

What makes the primitive children invalid? I assumed that React creates text nodes for mixed children, i.e.
```jsx
createElement('div', null, [createElement('br'), 'A'])
```

> support creating elements inside a fragment without wrapping it first

This sounds like my original goal of flattening nested Fragments. Looking at the debugger, I see that non-element children have the string value on the `props` and `__v` keys of the Preact vnode (the `__e` property also appears to indicate that it is a text node):
image

What would you think about using these keys to determine whether a child is a text node and returning that text alongside element siblings? Something like this for non-root Fragments:
```jsx
      return renderedChildren.map((child) => {
        if (typeof child.props === 'string') {
          return child.props;
        }
        return parseRenderedTree(child, child.__k);
      }) as VNode[];
```

> Another option might be to say "we don't support invalid markup". Maybe we shouldn't support this in the first place!

I don't know that I'm prepared to take on the responsibility of trying to make sense of markup 😛. Wherever possible we've tried to take the same approach as (P)React, primarily to ensure compatibility and not violate the principle of least surprise.

File: packages/container/src/render.ts [hunk 1]
```diff
@@ -140,11 +141,21 @@ export const composeRenderMethods: ComposeRenderMethodsCallback = ({
...
           fragmentChildren[0].__k
         );
-      }
+      } else {
+        // Handling for nested or non-root fragments. This will flatten non-root fragments
+
+        return renderedChildren.map((child) => {
+
+          // Return the text content directly if it's a text node
+          if (child.__e instanceof Text) {
```
Comment (andy-haynes) [inline packages/container/src/render.ts:150]: Checking for `Text` doesn't actually work here because `(Fragment instanceof Text) === true`. This Component fails to render because `parseRenderedTree` returns a `props` object instead of a DOM node:
```jsx
<>
  <>
    A<>B</>
  </>
</>
```

Since we need the Fragment to be parsed recursively, this check should only return the string leaf node. I think we do need the `props` string check after all 🙂

**Review by andy-haynes — CHANGES_REQUESTED** [review]
Just one thing otherwise LGTM 🎉

**Review by andy-haynes — APPROVED** [review]
❤️

File: packages/container/src/render.ts (hunk 1 above)
Comment (andy-haynes) [reply packages/container/src/render.ts]: Very cool, I'll take a look at those links.

> Out of curiosity, how'd you come up with that edge case? This code feels like a good candidate for some fuzz testing 👀

I just needed enough superfluous structure to really exercise the tree logic. It's actually trimmed down from a more pathological test Component 😂 
```jsx
    <>
      <>1</>
      <>
        <br />
        <>
          <input type="text" />
          <>
            A<>B</>
          </>
        </>
      </>
    </>
```

We definitely need some testing around this though, both on the rendered Components and individual modules.

Thanks again @ricky-sb for your help on this!
